google/cloud/spanner_dbapi/transaction_helper.py

```diff
@@ -0,0 +1,105 @@
+# Copyright 2023 Google LLC All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import TYPE_CHECKING, List
+
+import time
+from google.api_core.exceptions import Aborted
+
+from google.cloud.spanner_dbapi.utils import PeekIterator
+
+if TYPE_CHECKING:
+    from google.cloud.spanner_dbapi import Connection
+from google.cloud.spanner_dbapi.checksum import ResultsChecksum, _compare_checksums
+from google.cloud.spanner_dbapi.parsed_statement import Statement
+from google.cloud.spanner_v1.session import _get_retry_delay
+
+from google.rpc.code_pb2 import ABORTED
+
+MAX_INTERNAL_RETRIES = 50
+
+
+class TransactionHelper:
+    def __init__(self, connection: "Connection"):
+        self._connection = connection
+        # Non-Batch statements, which were executed within the current
+        # transaction
+        self._single_statements: List[Statement] = []
+        # Batch statements, which were executed within the current transaction
+        self._batch_statements_list: List[(List[Statement], ResultsChecksum)] = []
+
+    def retry_transaction(self):
+        """Retry the aborted transaction.
+
+        All the statements executed in the original transaction
+        will be re-executed in new one. Results checksums of the
+        original statements and the retried ones will be compared.
+
+        :raises: :class:`google.cloud.spanner_dbapi.exceptions.RetryAborted`
+            If results checksum of the retried statement is
+            not equal to the checksum of the original one.
+        """
+        attempt = 0
+        while True:
+            self._connection._spanner_transaction_started = False
+            attempt += 1
+            if attempt > MAX_INTERNAL_RETRIES:
+                raise
+
+            try:
+                self._rerun_previous_statements()
+                self._rerun_previous_batch_statements()
```

Comment:
This means that the retry will always first run all 'single' statements and then all batched statements. That can be a completely different order than in the original transaction, which makes the retry do something different than the original attempt.

Consider for example if the original attempt executed this:

```sql
start batch dml;
insert into foo (id) values (1);
insert into foo (id) values (2);
run batch;
select * from foo;
```

The retry will execute:

```
select * from foo;
start batch dml;
insert into foo (id) values (1);
insert into foo (id) values (2);
run batch;
```

That would cause the checksum of `select * from foo` to be different. The fact that no test cases seem to fail indicates that we do not have enough test coverage, as this should then certainly have failed.

Reply:
Ohh didn't thought about the order, thanks for catching this. Let me try to fix this